lib/lru_redux/cache.rb

```diff
@@ -1,104 +1,75 @@
-
-
+# Ruby 1.9 makes our life easier, Hash is already ordered
+#
+# This is an ultra efficient 1.9 freindly implementation
 class LruRedux::Cache
-
-  # for high efficiency nodes in double linked list are stored in arrays
-  # [prev,key,val,next]
-  #
-  # This makes this much more annoying to code, but gives us a 5-10% edge
-
   def initialize(max_size)
     @max_size = max_size
     @data = {}
-    @head = nil
-    @tail = nil
   end
 
   def max_size=(size)
     raise ArgumentError.new(:max_size) if @max_size < 1
     @max_size = size
-    while pop_tail
-      # no op
-    end
+
+    @data.shift while @data.size > @max_size
   end
 
   def getset(key)
-    node = @data[key]
-    if node
-      move_to_head(node)
-      node[2]
+    found = true
+    value = @data.delete(key){ found = false }
+    if found
+      @data[key] = value
     else
-      self[key] = yield
+      result = @data[key] = yield
+      @data.shift if @data.length > @max_size
+      result
     end
   end
 
   def fetch(key)
-    node = @data[key]
-    if node
-      move_to_head(node)
-      node[2]
+    found = true
+    value = @data.delete(key){ found = false }
+    if found
+      @data[key] = value
     else
       yield if block_given?
     end
   end
 
   def [](key)
-    node = @data[key]
-    if node
-      move_to_head(node)
-      node[2]
+    found = true
+    value = @data.delete(key){ found = false }
+    if found
+      @data[key] = value
+    else
+      nil
     end
   end
 
   def []=(key,val)
-    node = @data[key]
-    if node
-      move_to_head(node)
-      node[2] = val
-    else
-      @data[key] = add_to_head(key,val)
-      pop_tail
-    end
+    @data.delete(key)
+    @data[key] = val
+    @data.shift if @data.length > @max_size
     val
   end
 
   def each
-    a = to_a
-    a.each do |pair|
+    array = @data.to_a
+    array.reverse!.each do |pair|
```

Comment:
Could this use [`Hash#reverse_each`](https://ruby-doc.org/core-2.3.4/Enumerable.html#method-i-reverse_each) instead?

```ruby
  def each(*args, &block)
    @data.reverse_each(*args, &block)
  end
```